src/ucs/stats/stats.c

```diff
@@ -460,6 +465,138 @@ void ucs_stats_node_free(ucs_stats_node_t *node)
     }
 }
 
+static inline uint32_t ucs_stats_class_id_to_aggregate_sum_base_offset_get(
+        const char *class_name, unsigned *class_id,
+        size_t num_counters_in_class)
+{
+    uint32_t offset = 0;
+
+    if (ucs_likely(*class_id != 0)) {
+        offset = g_stats_class_id_to_name[*class_id].aggregate_sum_class_offset;
+    } else {
+        if (g_stats_class_id_to_name_size == MAX_NUM_OF_STATS_CLASSES) {
+            printf("%s:%d Warning: Statistics class_id to name array size too "
+                   "large!\n",
+                   __func__, __LINE__);
+            return 0;
+        }
+
+        /* Initialize entry */
+        snprintf(g_stats_class_id_to_name[g_stats_class_id_to_name_size]
+                         .class_name,
+                 sizeof(g_stats_class_id_to_name[g_stats_class_id_to_name_size]
+                                .class_name),
+                 "%s", class_name);
+        g_stats_class_id_to_name[g_stats_class_id_to_name_size]
+                .aggregate_sum_class_offset =
+                g_stats_class_id_to_name_last_offset;
+
+        /* Initialize return value */
+        offset = g_stats_class_id_to_name_last_offset;
+
+        /* Return class_id */
+        *class_id = g_stats_class_id_to_name_size;
```

Comment:
Looks like the function does two things:
1. allocate/assign new class_id-s for classes
2. return the offset for existing class_id-s

I suggest it would be much more efficient to assign class_id-s once, during ucs_stats_class creation, and during summation only use existing class_ids. This means splitting this function into these two cases, but also this way you'll also know the size of the array in advance.

Reply:
Agreed, as we discussed - The function will return the pointer to the const list of class_id's to name and not assign the names each time.

Reply:
Fixed.